src/service/history-impl.js

```diff
@@ -66,47 +81,71 @@ export class History {
      * }>} */
     this.queue_ = [];
 
-    this.binding_.setOnStackIndexUpdated(this.onStackIndexUpdated_.bind(this));
+    this.binding_.setOnStateUpdated(this.onStateUpdated_.bind(this));
   }
 
-  /** @private */
-  cleanup_() {
-    this.binding_.cleanup_();
+  /** @visibleForTesting */
+  cleanup() {
+    this.binding_.cleanup();
   }
 
   /**
    * Pushes new state into history stack with an optional callback to be called
-   * when this state is popped.
+   * when this state is popped as well as an object with updates to be applied
+   * to the state.
    * @param {!Function=} opt_onPop
+   * @param {!HistoryStateUpdateDef=} opt_stateUpdate
    * @return {!Promise<!HistoryIdDef>}
    */
-  push(opt_onPop) {
+  push(opt_onPop, opt_stateUpdate) {
     return this.enque_(() => {
-      return this.binding_.push().then(stackIndex => {
-        this.onStackIndexUpdated_(stackIndex);
+      return this.binding_.push(opt_stateUpdate).then(historyState => {
+        this.onStateUpdated_(historyState);
         if (opt_onPop) {
-          this.stackOnPop_[stackIndex] = opt_onPop;
+          this.stackOnPop_[historyState.stackIndex] = opt_onPop;
         }
-        return stackIndex;
+        return historyState.stackIndex;
       });
     }, 'push');
   }
 
   /**
    * Pops a previously pushed state from the history stack. If onPop callback
-   * has been registered, it will be called. All states coming after this
-   * state will also be popped and their callbacks executed.
-   * @param {!HistoryIdDef} stateId
+   * has been registered, it will be called with the state that was associated
+   * with the new head state within the history stack. All states coming
+   * after the supplied state will also be popped, and their
+   * callbacks executed in the same fashion.
+   * @param {!HistoryIdDef} stackIndex
    * @return {!Promise}
    */
-  pop(stateId) {
+  pop(stackIndex) {
```

Comment:
Let's keep this as `stateId`. It's just a mask for `stackIndex` but avoids leaking the implementation semantics to callers.